src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -0,0 +1,155 @@
+        /// <param name="handler">The handler that will handle the request.</param>
+        /// <param name="request">The request to handle.</param>
+        /// <param name="clientCapabilities">The client capabilities.</param>
+        /// <param name="clientName">The client name.</param>
+        /// <param name="requestCancellationToken">A cancellation token that will cancel the handing of this request.
+        /// The request could also be cancelled by the queue shutting down.</param>
+        /// <returns>A task that can be awaited to observe the results of the handing of this request.</returns>
+        public Task<TResponseType> ExecuteAsync<TRequestType, TResponseType>(
+            bool mutatesSolutionState,
+            IRequestHandler<TRequestType, TResponseType> handler,
+            TRequestType request,
+            ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken requestCancellationToken) where TRequestType : class
+        {
+            // Create a task completion source that will represent the processing of this request to the caller
+            var completion = new TaskCompletionSource<TResponseType>();
+
+            var item = new QueueItem(mutatesSolutionState, clientCapabilities, clientName,
+                callback: async (context, cancellationToken) =>
+                {
+                    // Check if cancellation was requested while this was waiting in the queue
+                    if (cancellationToken.IsCancellationRequested)
+                    {
+                        completion.SetCanceled();
+
+                        // Tell the queue to ignore any mutations from this request, not that we've given it a chance
+                        // to make any
+                        return false;
+                    }
+
+                    try
+                    {
+                        var result = await handler.HandleRequestAsync(request, context, cancellationToken).ConfigureAwait(false);
+                        completion.SetResult(result);
+                        // Tell the queue that this was successful so that mutations (if any) can be applied
+                        return true;
+                    }
+                    catch (OperationCanceledException)
+                    {
+                        completion.SetCanceled();
+                    }
+                    catch (Exception exception)
+                    {
+                        // Pass the exception to the task completion source, so the caller of the ExecuteAsync method can observe but
+                        // don't let it escape from this callback, so it doesn't affect the queue processing.
+                        completion.SetException(exception);
+                    }
+
+                    // Tell the queue to ignore any mutations from this request
+                    return false;
+                }, requestCancellationToken);
+            _queue.Enqueue(item);
```

Comment:
is asyncqueue threadsafe?

Reply:
per the docs it appears to be 
https://docs.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.asyncqueue-1?view=visualstudiosdk-2019

Reply:
Says so on the tin :)
https://github.com/microsoft/vs-threading/blob/master/src/Microsoft.VisualStudio.Threading/AsyncQueue.cs#L19